receiver/m365receiver/logs.go

```diff
@@ -0,0 +1,537 @@
+				l.logger.Error("unable to retrieve logs", zap.Error(err))
+				l.mu.Unlock()
+				return
+			}
+		} else {
+			l.logger.Error("error retrieving logs", zap.Error(err))
+			l.logger.Error(audit.route) //debug only
+			l.mu.Unlock()
+			return
+		}
+	}
+	l.mu.Unlock()
+
+	logs := l.transformLogs(pcommon.NewTimestampFromTime(now), audit, data)
+
+	l.consumerMu.Lock()
+	if logs.LogRecordCount() > 0 {
+		if err = l.consumer.ConsumeLogs(ctx, logs); err != nil {
+			l.logger.Error("error consuming events", zap.Error(err))
+		}
+	}
+	l.consumerMu.Unlock()
+}
+
+// constructs logs from logData
+func (l *m365LogsReceiver) transformLogs(now pcommon.Timestamp, audit *auditMetaData, data logData) plog.Logs {
+	logs := plog.NewLogs()
+	resourceLogs := logs.ResourceLogs().AppendEmpty()
+	ra := resourceLogs.Resource().Attributes()
+	ra.PutStr("m365.audit", audit.name)
+	ra.PutStr("m365.organization_id", l.cfg.TenantID)
+
+	for i := 0; i < len(data.logs); i++ {
+		log := data.logs[i]
+		logRecord := resourceLogs.ScopeLogs().AppendEmpty().LogRecords().AppendEmpty()
+
+		// log body
+		logRecord.Body().SetStr("\"C" + data.body[i])
+
+		// timestamp
+		ts, err := time.Parse(layout, log.CreationTime)
+		if err != nil {
+			l.logger.Warn("unable to interpret when an event was created, expecting a RFC3339 timestamp", zap.String("timestamp", log.CreationTime), zap.String("log", log.ID))
+			logRecord.SetTimestamp(now)
+		} else {
+			logRecord.SetTimestamp(pcommon.NewTimestampFromTime(ts))
+		}
+		logRecord.SetObservedTimestamp(now)
+
+		// attributes
+		attrs := logRecord.Attributes()
+		attrs.PutStr("id", log.ID)
+		attrs.PutStr("operation", log.Operation)
+		attrs.PutStr("user_id", log.UserID)
+		attrs.PutStr("user_type", matchUserType(log.UserType))
+
+		parseOptionalAttributes(&attrs, &log)
+	}
+
+	return logs
+}
+
+// sets the checkpoint
+func (l *m365LogsReceiver) checkpoint(ctx context.Context) error {
+	bytes, err := json.Marshal(l.record)
+	if err != nil {
+		return fmt.Errorf("unable to write checkpoint: %w", err)
+	}
+	return l.storageClient.Set(ctx, logStorageKey, bytes)
+}
+
+// loads the checkpoint
+func (l *m365LogsReceiver) loadCheckpoint(ctx context.Context) {
+	bytes, err := l.storageClient.Get(ctx, logStorageKey)
+	if err != nil {
+		l.logger.Info("unable to load checkpoint from storage client, continuing without a previous checkpoint", zap.Error(err))
+		l.record = &logRecord{}
+		return
+	}
+
+	if bytes == nil {
+		l.record = &logRecord{}
+		return
+	}
+
+	var record logRecord
+	if err = json.Unmarshal(bytes, &record); err != nil {
+		l.logger.Error("unable to decode stored record for events, continuing without a checkpoint", zap.Error(err))
+		l.record = &logRecord{}
+		return
+	}
+	l.record = &record
+}
+
+// adds any attributes to log that may or may not be present
+func parseOptionalAttributes(m *pcommon.Map, log *jsonLogs) {
+	recordType := matchRecordType(log.RecordType)
+	if log.Workload != "" {
```

Comment:
Repeated code setting simple string optional attributes can be refactored to a function, something like

```
func setAttributeIfNotEmpty(m *pcommon.Map, field, value string) {
    if field != "" {
        m.PutStr(field, value)
    }
}
```